sharktank/sharktank/utils/llm_utils.py

```diff
@@ -538,6 +391,83 @@ def decode(
         logits, indices = decode_task.run(*self._cache)
         return logits, indices
 
+    def submit_tasks(
+        self,
+        requests: list[list[int]],
+        page_ids: list[list[int]],
+        steps: int,
+    ):
+        assert len(requests) == len(page_ids)
+
+        for i, request in enumerate(requests):
+            task_input = LlmTaskInput(
+                task_id=f"req_{i}",
+                tokens=request,
+                seq_len=len(request),
+                pages=page_ids[i],
+            )
+            # Submit prefill task
+            self._prefill_scheduler.schedule_task(task_input, 1)
+            # Submit decode tasks
+            self._decode_scheduler.schedule_task(task_input, steps - 1)
+
+    def run(
+        self,
+        selection_fn: Callable[
+            [numpy.ndarray, Optional[numpy.ndarray], list[int]],
+            list[int],
+        ],
+        eos_token: int,
+        steps: int,
+    ):
+        selections_map = {}
+        input_keys = []
+        task_inputs = self._prefill_scheduler.get_scheduled_tasks()
+        for task in task_inputs:
+            selections_map[task.task_id] = []
+            input_keys.append(task.task_id)
+
+        # Run prefill
+        prefill_task = PrefillTask(
+            invocation_fn=self._instance.prefill,
+            llm_task_inputs=task_inputs,
+            batch_size=self._prefill_bs,
+            block_stride=self._block_stride,
+        )
+        logits, indices = prefill_task.run(*self._cache)
+        last = selection_fn(
+            logits,
+            indices,
+            [task_inputs.seq_len - 1 for task_inputs in task_inputs],
+        )
+        for task_input, token in zip(task_inputs, last):
+            selections_map[task_input.task_id].append(token)
+            self._prefill_scheduler.on_task_complete(task_input.task_id, token)
+
+        # Run decode steps
+        for _ in range(steps - 1):
+            if not self._decode_scheduler.has_pending_tasks():
+                break
+            task_inputs = self._decode_scheduler.get_scheduled_tasks()
+            decode_task = DecodeTask(
+                invocation_fn=self._instance.decode,
+                llm_task_inputs=task_inputs,
+                batch_size=self._decode_bs,
+                block_stride=self._block_stride,
+            )
+            logits, indices = decode_task.run(*self._cache)
```

Comment:
This looks wrong.  You make a task then call run on it then  call the scheduler? You should be making a job and submitting it. The scheduler should basically have a `schedule` command for each task (a task being a separate request). Then a call for competion. Something is off on this.